### PR TITLE
Refined types as per `laminas/laminas-coding-standard:2.3.x` upgrades

### PR DESCRIPTION
**Pull request by Ocramius**
In https://github.com/laminas/laminas-diactoros/pull/104 (duplicate), we improved the types of internal and
public API. These were mixed with coding style improvements.

The coding style improvements have been merged separately via https://github.com/laminas/laminas-diactoros/pull/103,
so this patch only includes the type/docblock improvements, which should ease the potential evolution of the
API of this package.

|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | yes

### CODE REVIEW

**Comment by Ocramius** [comment]
@gsteel you may be interested in checking this one - it's very closely related to your #104 work, but with some type refinements

File: src/Exception/InvalidStreamPointerPositionException.php [hunk 1]
```diff
@@ -9,10 +9,10 @@
 
 class InvalidStreamPointerPositionException extends RuntimeException implements ExceptionInterface
 {
-    /** @param int $code */
+    /** {@inheritDoc} */
     public function __construct(
         string $message = 'Invalid pointer position',
-        $code = 0,
+        int $code = 0,
```
Comment (Ocramius) [inline src/Exception/InvalidStreamPointerPositionException.php:15]: Non-BC break, as it is part of the constructor. Should a consumer pass a non-`int` here, due to `declare(strict_types=1)`, a crash would be happening during the `parent::__construct()` call anyway

File: src/Exception/UploadedFileAlreadyMovedException.php [hunk 2]
```diff
@@ -9,10 +9,10 @@
 
 class UploadedFileAlreadyMovedException extends RuntimeException implements ExceptionInterface
 {
-    /** @param int $code */
+    /** {@inheritDoc} */
     public function __construct(
         string $message = 'Cannot retrieve stream after it has already moved',
-        $code = 0,
+        int $code = 0,
```
Comment (Ocramius) [inline src/Exception/UploadedFileAlreadyMovedException.php:15]: Non-BC break, as it is part of the constructor. Should a consumer pass a non-`int` here, due to `declare(strict_types=1)`, a crash would be happening during the `parent::__construct()` call anyway

File: test/HeaderSecurityTest.php [hunk 3]
```diff
@@ -34,42 +34,47 @@ public function getFilterValues(): array
...
+            ["This is a \r\n\n\r\n test", false],
+            ["This is a \xFF test", false],
+            ["This is a \x7F test", false],
+            ["This is a \x7E test", true],
         ];
     }
 
     /**
      * @dataProvider validateValues
+     * @group ZF2015-04
```
Comment (Ocramius) [inline test/HeaderSecurityTest.php:69]: Note: I re-introduced the various `@group` annotations.

While they may look outdated, they are really useful to dig up the reason why changes happened, especially after we moved code around repositories/subtree splits/etc.

Comment (gsteel) [reply test/HeaderSecurityTest.php:69]: I'll remember that for next time 👍

File: src/Exception/UploadedFileAlreadyMovedException.php (hunk 2 above)
Comment (boesing) [reply src/Exception/UploadedFileAlreadyMovedException.php:15]: ~We added `declare(strict_types=1)` in one of the latest minor. Wasn't that already a BC break then?~

nvm, it was there prior applying latest codestyle

File: src/Request/ArraySerializer.php [hunk 4]
```diff
@@ -23,6 +23,15 @@ final class ArraySerializer
 {
     /**
      * Serialize a request message to an array.
+     *
+     * @return array{
```
Comment (boesing) [inline src/Request/ArraySerializer.php:27]: Do we use this kind of array somewhere?
If so, I would prefer having a `psalm-type` so we can import that array shape for further usage.

File: src/Response/ArraySerializer.php [hunk 5]
```diff
@@ -23,6 +23,14 @@ final class ArraySerializer
 {
     /**
      * Serialize a response message to an array.
+     *
+     * @return array{
```
Comment (boesing) [inline src/Response/ArraySerializer.php:27]: Do we use this kind of array somewhere?
If so, I would prefer having a psalm-type so we can import that array shape for further usage.

File: src/functions/parse_cookie_header.php [hunk 6]
```diff
@@ -16,7 +16,7 @@
  * overwriting. Thus, the server request should take the cookies from the request header instead.
  *
  * @param string $cookieHeader A string cookie header value.
- * @return array key/value cookie pairs.
+ * @return array<string, string> key/value cookie pairs.
```
Comment (boesing) [inline src/functions/parse_cookie_header.php]: can we be more strict here? guess its `array<non-empty-string,string>`?
I can't beliefe that a cookie header like this is supported:

```
Cookie: =
```

File: src/functions/parse_cookie_header.legacy.php [hunk 7]
```diff
@@ -8,9 +8,10 @@
 use function Laminas\Diactoros\parseCookieHeader as laminas_parseCookieHeader;
```
Comment (boesing) [inline src/functions/parse_cookie_header.legacy.php:8]: Maybe we just drop this as we already removed support for the old zendframework implementation by adding `conflict` to `composer.json`?

**Review by boesing — CHANGES_REQUESTED** [review]

File: src/Request/ArraySerializer.php (hunk 4 above)
Comment (Ocramius) [reply src/Request/ArraySerializer.php:27]: We don't currently use this

File: src/Response/ArraySerializer.php (hunk 5 above)
Comment (Ocramius) [reply src/Response/ArraySerializer.php:27]: We don't currently use this

File: src/functions/parse_cookie_header.php (hunk 6 above)
Comment (Ocramius) [reply src/functions/parse_cookie_header.php]: Doesn't seem to be supported: will add `non-empty-string` here, although I cannot safely add an assertion, and will expand the baseline instead

File: src/functions/parse_cookie_header.legacy.php (hunk 7 above)
Comment (Ocramius) [reply src/functions/parse_cookie_header.legacy.php:8]: We do declare these symbols: we can drop them, but not in this patch :thinking:

File: src/functions/parse_cookie_header.php (hunk 6 above)
Comment (Ocramius) [reply src/functions/parse_cookie_header.php]: dfa990d294c2e6300699978e78c63724647c5d2a

**Review by gsteel — APPROVED** [review]

**Comment by Ocramius** [comment]
:ship: then - will cut a release, hoping for quick feedback in case of breakages

**Pull request merged**